Add player functionality.

mkulikiewicz opened the pull request:
Now two player can play in the game. They symbole are changed.

kulikiewiczm committed:
Add player functionality.

Now two player can play in the game. They symbole are changed.

patrycja93 commented on src/main/java/com/mkul/game/ox/NotEnoughPlayers.java:
```diff
@@ -0,0 +1,7 @@
+public class NotEnoughPlayers extends Exception {
+    NotEnoughPlayers() {
+        super("Not enough players in game");
+    }
```
Czasem wypisujesz komunikaty w języku angielskim jak wyżej, a czasem w polskim. Może warto to ujednolicić :)

patrycja93 commented on src/main/java/com/mkul/game/ox/OXApp.java:
```diff
@@ -8,10 +8,21 @@
+        Players players = new Players();
+        players.addPlayer("Maciek");
+        players.addPlayer("Michał");
+
```
Metoda addPlayer() nie musi być publiczna w klasie Players.

patrycja93 replied:
Masz jedna niepotrzebną pustą linie :smile:

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+
+    public void addPlayer(String name) throws TooManyPlayers {
+        new CreatePlayer().withName(name).withRandomFieldValue().addToList();
+    }
```
javadoc

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+    public FieldValue getFieldValueForCurrentPlayer() throws NotEnoughPlayers {
+        final Player currentPlayer = getCurrent();
+        return currentPlayer.getFieldValue();
+    }
```
javadoc

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+            return this.currentPlayer;
+        }
+        throw new NotEnoughPlayers();
+    }
```
javadoc

patrycja93 commented on src/main/java/com/mkul/game/ox/OXApp.java:
```diff
@@ -24,9 +35,9 @@ private static void run(String[] args) {
         }
```
Kod który się potwarza, warto w jakiś sposób wyekstrahować do na przykład metody . Tutaj masz powtórzenie w liniach 29-30 i 33-34. Może jakaś metoda przyjmująca rozmiar dla board i dodać ją w bloku finally po catch ? Dzięki temu zmniejszysz linie kodu i nie będzie powtórzeń :laughing:

AleksanderStudnicki commented on src/test/java/com/mkul/game/ox/PlayersTest.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Test
+    public void testGetCurrentPlayerReturnRandomPlayer() throws NotEnoughPlayers {
+        //Given
```
If you want to keep GWT structure maybe add some comment after given (something like "Given - before method"). It is only suggestion because I write something like that in "Then" when I want to test exception and do not have any code in then section (I noticed you write additional commentary in that case).

AleksanderStudnicki commented on src/test/java/com/mkul/game/ox/PlayersTest.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Test
+    public void testGetNextPlayerChangeCurrentPlayer() throws NotEnoughPlayers {
+        //Given
```
Same as previous.

AleksanderStudnicki commented on src/test/java/com/mkul/game/ox/PlayersTest.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test
+    public void testGetCurrentPlayerReturnFirstPlayer() throws NotEnoughPlayers, TooManyPlayers {
```
Same as previous but here without "Given" comment.

AleksanderStudnicki commented on src/test/java/com/mkul/game/ox/PlayersTest.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test(expectedExceptions = NotEnoughPlayers.class)
+    public void testGetNextPlayerThrowNotEnoughtPlayersException() throws NotEnoughPlayers, TooManyPlayers {
```
Same as previous.

patrycja93 replied:
To samo co wyżej, brak sekcji given :mag:

AleksanderStudnicki commented:
Consider using the checkstyle plugin. I pushed review branch "reviewAlek" with checkstyle in pom.xml, so run mvn site and look at the reports.

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Player.java:
```diff
@@ -0,0 +1,21 @@
+class Player {
+
+    private String name;
+    private FieldValue fieldValue;
```
Can be final.

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+        throw new NotEnoughPlayers();
+    }
+
+    private class CreatePlayer {
```
You can call it just "builder". When I see something like "create" or "build" I'm thinking about a method not a class. So just call it (Player)Builder or (Player)Creator.

AleksanderStudnicki commented on src/test/java/com/mkul/game/ox/boardcheck/BoardCheckerTest.java:
```diff
@@ -4,19 +4,18 @@
 public class BoardCheckerTest {
 
     private static BoardScore boardScore;
     private static final int WINING_SIZE = 3;
```
WINNING not WINING

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+
+    public FieldValue getFieldValueForCurrentPlayer() throws NotEnoughPlayers {
+        final Player currentPlayer = getCurrent();
+        return currentPlayer.getFieldValue();
```
Why final? Could be on one line getCurrent().getFieldValue().

patrycja93 commented on src/main/java/com/mkul/game/ox/OXApp.java:
```diff
@@ -47,13 +58,7 @@ private static void run(String[] args) {
-            }
+            players.getNext();
         }
     }
```
Ten while jest za długi. Może dobrze by było rozbić to na mniejsze metody. Jedna, zajmuje się pobieraniem danych od użytkownika, pobiera x i y, dopóki nie są poprawne i mogłaby zwracać obiekt typu Pair. Druga pobierałaby Pair i sprawdzała czy któryś z graczy wygrał :+1:  

Ogołnie metoda jest za długa i robi za dużo rzeczy(wiem, ze o tym wiesz). Przemyśl przeniesienie jej do nowej klasy i rozdzielenie na minimum 3 mniejsze metody.

patrycja93 commented on src/main/java/com/mkul/game/ox/Player.java:
```diff
@@ -0,0 +1,21 @@
+
+    private String name;
+    private FieldValue fieldValue;
+
```
FieldValue jest nieintuicyjne, może lepiej zmienić to na sign, symbol, mark ... :smiley:

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+                throw new TooManyPlayers();
+            }
+            playerList.add(new Player(name, fieldValue));
+            return playerList.peekLast();
```
I think you should just build new player and add to list outside builder.

patrycja93 commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+
+class Players {
+    private LinkedList<Player> playerList = new LinkedList<>();
+    private Player currentPlayer;
```
Dlaczego używasz klasę  implementującą interface List? W przypadku gdy będziesz chciał zmieniać tą implementacje na inną będzie to skutkowało dużą ilością zmian w kodzie :P

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+
+class Players {
```
I think you should create builder for Players not for a Player inside Players. Players should only exists when there are 2 of them. You can create players outside Players class and then pass them to Players builder. Also I think you can delete players list and change it to field for player1 and field for player2 (or something else, I'm not sure cause I don't implement it in my project yet).

AleksanderStudnicki replied:
I think you should take a look at that class and rebuild it.

patrycja93 commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private class CreatePlayer {
+        private String name;
```
Może lepie usunąć tą klase i przesieść te metody do Player ? Do rozważenia :grinning:

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,73 @@
+
+    public Player getNext() throws NotEnoughPlayers {
+        final Player currentPlayer = getCurrent();
+        Optional<Player> expectedNextPlayer = playerList.stream().filter(player -> player.getFieldValue() != currentPlayer.getFieldValue()).findFirst();
```
Too much in one line.

AleksanderStudnicki reviewed (commented):
I don't comment on demo class cause you know what to do. I wrote my comments elsewhere, so check them and fix what is necessary.

patrycja93 commented on src/test/java/com/mkul/game/ox/PlayersTest.java:
```diff
@@ -0,0 +1,102 @@
+        multiplePlayers.getCurrent();
+        //Then
+        assertThat(multiplePlayers.getCurrent().toString()).contains(" posiada znak ");
+        assertThat(multiplePlayers.getCurrent().toString())
```
Zmienna " posiada znak " można dodać do sekcji given :D

patrycja93 commented on src/test/java/com/mkul/game/ox/PlayersTest.java:
```diff
@@ -0,0 +1,102 @@
+
+        //Then
+        assertThat(secondPlayer.toString()).isEqualTo(multiplePlayers.getCurrent().toString());
+    }
```
To samo co wyżej, multiplePlayers.getCurrent().toString() można dodać do sekcji given. :smile:

patrycja93 commented on src/test/java/com/mkul/game/ox/PlayersTest.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Test
+    public void testGetCurrentPlayerReturnFirstPlayer() throws NotEnoughPlayers, TooManyPlayers {
+        //When
```
Brak sekcji given :dancer:

patrycja93 commented on src/test/java/com/mkul/game/ox/boardcheck/BoardCheckerTest.java:
```diff
@@ -4,19 +4,18 @@
 
     private static BoardScore boardScore;
     private static final int WINING_SIZE = 3;
 
```
WINNING *

kulikiewiczm committed:
Add checkstyle with google checker

Adapting the code to check style requirements

kulikiewiczm committed:
Rebuild adding player concept

Actualy players are adding in builder.
Players store only list of player and can move to next player.

kulikiewiczm committed:
Adding a format like in checkstyle

foramt alse added to tests classes

kulikiewiczm committed:
refactor PlayersTest class

kulikiewiczm committed:
Add spotbugs version to properties

AleksanderStudnicki commented on pom.xml:
```diff
@@ -65,6 +67,21 @@
                     </reportSet>
                 </reportSets>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-checkstyle-plugin</artifactId>
+                <version>${version.plugin.maven.checkstyle}</version>
+                <configuration>
+                    <configLocation>google_checks.xml</configLocation>
+                    <failOnViolation>false</failOnViolation>
+                    <enableFilesSummary>false</enableFilesSummary>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>com.github.spotbugs</groupId>
+                <artifactId>spotbugs-maven-plugin</artifactId>
+                <version>${version.spotbugs.maven.plugin}</version>
+            </plugin>
```
Noticable improvement :)

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Sign.java:
```diff
@@ -0,0 +1,20 @@
+ * Represent value of field in board.
+ */
+public enum Sign {
+  X("X"),
+  O("O");
```
String values are redundant, so delete them or change enum names (X -> CROSS, O -> NAUGHT).

AleksanderStudnicki commented on src/main/java/com/mkul/game/ox/Players.java:
```diff
@@ -0,0 +1,56 @@
+    return players.peek();
+  }
+
+  Player getNext() {
+    Player player = players.poll();
+    players.add(player);
+    return players.peek();
+  }
```
Method name does not tell what it really does ((poll -> add) -> peek). In the previous method there was also "get" prefix but method did only peek.

patrycja93 commented on src/main/java/com/mkul/game/ox/App.java:
```diff
@@ -0,0 +1,75 @@
+      return false;
+    }
+  }
+}
```
Sprawdzanie danych pobieranych od użytkownika mogłoby być w zupełnie innej klasie :)

patrycja93 commented on src/main/java/com/mkul/game/ox/BoardDraw.java:
```diff
@@ -2,41 +2,43 @@
+      System.out.printf("%n");
     }
+  }
 
```
Może warto byłoby stworzyć klase no UI ktora zajowałaby się pobieraniem dane od uzytkownika i wyswietlala wiadomosci.
